tensorflow/core/kernels/data/shard_dataset_op.cc

```diff
@@ -128,21 +128,33 @@ class ShardDatasetOp::Dataset : public DatasetBase {
         return Status::OK();
       }
 
-      std::vector<Tensor> result;
-      do {
-        result.clear();
-        TF_RETURN_IF_ERROR(input_impl_->GetNext(ctx, &result, end_of_sequence));
-        if (*end_of_sequence) {
-          input_impl_.reset();
-          return Status::OK();
-        }
-      } while ((next_index_++ % dataset()->num_shards_) != dataset()->index_);
+      int num_to_skip = (dataset()->index_ - next_index_) %
+                        dataset()->num_shards_;
+      if (num_to_skip < 0) {
+        num_to_skip += dataset()->num_shards_;
+      }
+      int num_skipped;
+      TF_RETURN_IF_ERROR(input_impl_->Skip(ctx, num_to_skip, end_of_sequence,
+                                           &num_skipped));
+      next_index_ += num_skipped;
```

Comment:
Do we need to keep track of `next_index_`? It seems like each call to `GetNext` could call `input_impl->Skip(num_shards - 1)`, then `input_impl_->GetNext()`, and there's no need to remember the index across calls to `GetNext`

Reply:
@aaudiber 
The `num_to_skip` has 3 possibilities:
- `dataset()->num_shards_ - 1` as you mentioned.
- `dataset()->index_` for the first time the iterator is called.
- `dataset()->index_` for the second time it is called when `dataset()->require_non_empty_` is true.

If we want to remove `next_index_`, I believe we need to add a member variable to distinguish these possibilities. Any thought on this?

Reply:
Got it, I missed the `require_non_empty_` handling. Your current implementation looks good to me.